File: src/readers/osm/way.ts
import { Info } from './info';
import { extendBBox } from '../../geometry';

import type { Metadata } from './primitive';
import type { PbfReader } from 'pbf-ts';
import type { PrimitiveBlock } from './primitive';
import type { InfoBlock, OSMProperties, OSMReader } from '.';

import type {
  BBOX,
  Properties,
  VectorFeature,
  VectorGeometry,
  VectorLineString,
} from '../../geometry';

/** Linebased node reference store */
export type WayNodes = number[];

/** An intermediate vector feature where the way nodes haven't been resolved yet. */
export interface IntermediateWay {
  id: number;
  properties: OSMProperties;
  metadata: InfoBlock;
  wayNodes: WayNodes;
  isArea: boolean;
}

/**
 * Convert an intermediate way to a vector feature
 * @param way - intermediate way
 * @param reader - OSM reader
 * @returns - way vector feature
 */
export async function intermediateWayToVectorFeature(
  way: IntermediateWay,
  reader: OSMReader,
): Promise<VectorFeature<Metadata, Properties, OSMProperties> | undefined> {
  const { addBBox } = reader;
  const { id, isArea, wayNodes, properties, metadata } = way;
  // build line
  const vectorLine: VectorLineString = [];
  for (const ref of wayNodes) {
    const node = await reader.nodeGeometry.get(ref);
    if (node === undefined) return;
    vectorLine.push({ ...node });
  }
  let bbox: BBOX | undefined;
  if (addBBox) {
    for (const node of vectorLine) bbox = extendBBox(bbox, node);
  }
  // build geometry
  const geometry: VectorGeometry = isArea
    ? { type: 'Polygon', is3D: false, coordinates: [vectorLine], bbox }
    : { type: 'LineString', is3D: false, coordinates: vectorLine, bbox };

  return {
    id,
    type: 'VectorFeature',
    properties,
    geometry,
    metadata: { type: 'way', info: metadata },
  };
}

/**
 * Way Class
 */
export class Way {
  id = -1;
  info?: Info;
  // Parallel arrays
  #keys: number[] = [];
  #vals: number[] = [];
  #refs: number[] = []; // DELTA coded
  // Optional infield lat-lon
  // NOTE: I'm not going to bother implementing this, I've never seen it used.
  //   #lats: number[] = []; // optional DELTA coded
  //   #lons: number[] = []; // optional DELTA coded

  /**
   * @param primitiveBlock - the primitive block to access keys and values
   * @param reader - the OSM reader
   * @param pbf - the Protobuf object to read from
   */
  constructor(
    public primitiveBlock: PrimitiveBlock,
    public reader: OSMReader,
    pbf: PbfReader,
  ) {
    pbf.readMessage(this.#readLayer, this);
  }

  /**
   * Checks if the way is filterable or not
   * @returns - true if the way is filterable
   */
  isFilterable(): boolean {
    const { tagFilter, skipWays } = this.reader;
    if (skipWays) return true;
    if (tagFilter !== undefined) {
      for (let i = 0; i < this.#keys.length; i++) {
        const keyStr = this.primitiveBlock.getString(this.#keys[i]);
        const valStr = this.primitiveBlock.getString(this.#vals[i]);
        if (tagFilter.matchFound('Way', keyStr, valStr)) return false;
      }
      // if we make it here, we didn't find any matching tags
      return true;
    }
    return false;
  }

  /**
   * Access the way's properties
   * @returns - the way's properties
   */
  properties(): OSMProperties {
    return this.primitiveBlock.tags(this.#keys, this.#vals);
  }

  /**
   * Access the way's node IDs associated with this way
   * @returns - the way's nodes
   */
  nodeRefs(): number[] {
    const res: number[] = [];
    let ref = 0;
    for (let i = 0; i < this.#refs.length; i++) {
      ref += this.#refs[i];
      res.push(ref);
    }
    return res;
  }

  /**
   * Checks if the way has a key-value pair (value optional)
   * @param key - the key
   * @param val - the value
   * @returns - true if the way has the key and value
   */
  hasKeyValue(key: string, val?: string): boolean {
    const { primitiveBlock: pb } = this;
    for (let i = 0; i < this.#keys.length; i++) {
      if (pb.getString(this.#keys[i]) === key) {
        if (val === undefined) return true;
        if (pb.getString(this.#vals[i]) === val) return true;
      }
    }
    return false;
  }

  /**
   * Checks if the way is an area based on it's key-value pairs
   * @returns - true if the way is an area
   */
  isArea(): boolean {
    const { upgradeWaysToAreas } = this.reader;
    if (
      (upgradeWaysToAreas &&
        this.#refs.length >= 4 &&
        this.#refs[0] === this.#refs[this.#refs.length - 1]) ||
      this.hasKeyValue('area', 'yes')
    ) {
      return true;
    }
    return false;
  }

  /**
   * Converts the way to an intermediate vector feature (way's nodes have not been parsed)
   * @returns - the way as an intermediate vector feature
   */
  toVectorFeature(): undefined | IntermediateWay {
    const isArea = this.isArea();
    const wayNodes = this.nodeRefs();
    if (wayNodes.length < 2) return;
    return {
      id: this.id,
      isArea,
      properties: this.properties(),
      wayNodes,
      metadata: this.info?.toBlock() ?? {},
    };
  }

  /**
   * @param tag - the tag of the message
   * @param way - the way to modify
   * @param pbf - the Protobuf object to read from
   */
  #readLayer(tag: number, way: Way, pbf: PbfReader): void {
    if (tag === 1) way.id = pbf.readVarint();
    else if (tag === 2) way.#keys = pbf.readPackedVarint();
    else if (tag === 3) way.#vals = pbf.readPackedVarint();
    else if (tag === 4) way.info = new Info(way.primitiveBlock, pbf);
    else if (tag === 8) way.#refs = pbf.readPackedSVarint();
    // skip, not used.
    else if (tag === 9 || tag === 10) return;
    // else if (tag === 9) way.#lats = pbf.readPackedSVarint();
    // else if (tag === 10) way.#lons = pbf.readPackedSVarint();
    else throw new Error(`Unknown tag: ${tag}`);
  }
}

File: rust/readers/osm/way.rs
use super::{
    OSMFilterable, OSMReader, OSMTagFilterType,
    info::{Info, InfoBlock},
    node::IntermediateNode,
    primitive::{OSMMetadata, PrimitiveBlock},
    relation::{IntermediateRelation, MemberType},
};
use crate::{data_store::kv::KVStore, readers::Reader};
use alloc::{vec, vec::Vec};
use pbf::{ProtoRead, Protobuf};
use s2json::{
    BBox3D, Properties, VectorFeature, VectorFeatureType, VectorGeometry, VectorLineString,
    VectorPoint,
};
use serde::{Deserialize, Serialize};

/// Linebased node reference store
pub type WayNodes = Vec<u64>;

/// An intermediate vector feature where the way nodes haven't been resolved yet.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct IntermediateWay {
    /// The way's ID
    pub id: u64,
    /// The way's properties
    pub properties: Properties,
    /// Optional metadata
    pub info: Option<InfoBlock>,
    /// The way's nodes IDs to be resolved
    pub way_nodes: WayNodes,
    /// Whether the way is an area
    pub is_area: bool,
}
impl IntermediateWay {
    /// Convert the node to a vector feature
    pub fn to_vector_feature<_N: KVStore<u64, VectorPoint<()>>>(
        &self,
        node_geometry: &_N,
        add_bbox: bool,
    ) -> VectorFeature<OSMMetadata, Properties, ()> {
        let IntermediateWay { id, is_area, way_nodes, properties, info } = &self;
        let mut bbox = BBox3D::default();
        // build line
        let mut vector_line: VectorLineString<()> = vec![];
        for way_node in way_nodes {
            let node = node_geometry.get(*way_node);
            if let Some(node) = node {
                if add_bbox {
                    bbox.extend_from_point(node)
                }
                vector_line.push(node.clone());
            }
        }
        // build geometry
        let bbox = if add_bbox { Some(bbox) } else { None };
        let geometry = match is_area {
            true => VectorGeometry::new_polygon(vec![vector_line], bbox),
            false => VectorGeometry::new_linestring(vector_line, bbox),
        };

        VectorFeature {
            id: Some(*id),
            face: 0.into(),
            _type: VectorFeatureType::VectorFeature,
            properties: properties.clone(),
            geometry,
            metadata: Some(OSMMetadata {
                osm_type: MemberType::Way,
                info: info.clone(),
                nodes: None,
                relation: None,
            }),
        }
    }
}

/// Way Class
#[derive(Debug, Default)]
pub struct Way {
    /// The way's ID
    pub id: u64,
    info: Option<Info>,
    // Parallel arrays
    keys: Vec<u32>,
    vals: Vec<u32>,
    // DELTA coded
    refs: Vec<i64>,
    // Optional infield lat-lon
    // NOTE: I'm not going to bother implementing this, I've never seen it used.
    //   lats: Vec<i32>, // optional DELTA coded
    //   lons: Vec<i32>, // optional DELTA coded */
}
impl Way {
    /// Get the properties of the node
    pub fn properties(&self, pb: &PrimitiveBlock) -> Properties {
        pb.tags(&self.keys, &self.vals)
    }

    /// Checks if the way is an area based on it's key-value pairs
    pub fn is_area<
        T: Reader,
        _N: KVStore<u64, VectorPoint<()>>,
        N: KVStore<u64, IntermediateNode>,
        _W: KVStore<u64, WayNodes>,
        W: KVStore<u64, IntermediateWay>,
        R: KVStore<u64, IntermediateRelation>,
    >(
        &self,
        pb: &PrimitiveBlock,
        reader: &mut OSMReader<T, _N, N, _W, W, R>,
    ) -> bool {
        if (reader.upgrade_ways_to_areas
            && self.refs.len() >= 4
            && self.refs[0] == self.refs[self.refs.len() - 1])
            || self.has_key_value(pb, "area", Some("yes"))
        {
            return true;
        }
        false
    }

    /// Checks if the way has a key-value pair (value optional)
    pub fn has_key_value(&self, pb: &PrimitiveBlock, key: &str, val: Option<&str>) -> bool {
        for i in 0..self.keys.len() {
            if pb.get_string(self.keys[i] as usize) == key {
                match val {
                    None => return true,
                    Some(v) => {
                        if pb.get_string(self.vals[i] as usize) == v {
                            return true;
                        }
                    }
                }
            }
        }
        false
    }

    /// Access the way's node IDs associated with this way
    pub fn node_refs(&self) -> WayNodes {
        let mut res = vec![];
        let mut _ref = 0;
        // for (let i = 0; i < this.#refs.length; i++) {
        for i in 0..self.refs.len() {
            _ref += self.refs[i];
            res.push(_ref as u64);
        }
        res
    }

    /// Converts the way to an intermediate vector feature (way's nodes have not been parsed)
    /// @returns - the way as an intermediate vector feature
    pub fn to_intermediate_feature<
        T: Reader,
        _N: KVStore<u64, VectorPoint<()>>,
        N: KVStore<u64, IntermediateNode>,
        _W: KVStore<u64, WayNodes>,
        W: KVStore<u64, IntermediateWay>,
        R: KVStore<u64, IntermediateRelation>,
    >(
        &self,
        pb: &PrimitiveBlock,
        reader: &mut OSMReader<T, _N, N, _W, W, R>,
    ) -> Option<IntermediateWay> {
        let is_area = self.is_area(pb, reader);
        let way_nodes = self.node_refs();
        if way_nodes.len() < 2 {
            None
        } else {
            Some(IntermediateWay {
                id: self.id,
                is_area,
                properties: self.properties(pb),
                way_nodes,
                info: self.info.as_ref().map(|info| info.to_block(pb)),
            })
        }
    }
}
impl OSMFilterable for Way {
    fn is_filterable<
        T: Reader,
        _N: KVStore<u64, VectorPoint<()>>,
        N: KVStore<u64, IntermediateNode>,
        _W: KVStore<u64, WayNodes>,
        W: KVStore<u64, IntermediateWay>,
        R: KVStore<u64, IntermediateRelation>,
    >(
        &self,
        pb: &PrimitiveBlock,
        reader: &mut OSMReader<T, _N, N, _W, W, R>,
    ) -> bool {
        if reader.skip_ways {
            return true;
        }
        if let Some(tag_filter) = &mut reader.tag_filter {
            for i in 0..self.keys.len() {
                let key_str = pb.get_string(self.keys[i] as usize);
                let val_str = pb.get_string(self.vals[i] as usize);
                if tag_filter.match_found(OSMTagFilterType::Way, key_str, val_str) {
                    return false;
                }
            }
            // if we make it here, we didn't find any matching tags
            return true;
        }
        false
    }
}
/// Read in the contents of the way
impl ProtoRead for Way {
    fn read(&mut self, tag: u64, pb: &mut Protobuf) {
        match tag {
            1 => self.id = pb.read_varint(),
            2 => self.keys = pb.read_packed(),
            3 => self.vals = pb.read_packed(),
            4 => {
                let mut info = Info::default();
                pb.read_message(&mut info);
                self.info = Some(info);
            }
            8 => self.refs = pb.read_s_packed(),
            // skip, not used.
            9 | 10 => (),
            _ => panic!("unknown tag {}", tag),
        }
    }
}
